Index: node_modules/epoll/epoll.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- node_modules/epoll/epoll.js	(date 1462305420000)
+++ node_modules/epoll/epoll.js	(date 1462305420000)
@@ -1,2 +1,2 @@
-module.exports = require('bindings')('epoll.node');
+module.exports = require('./build/Release/epoll');
 
