Index: node_modules/fastcall/lib/native.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- node_modules/fastcall/lib/native.js	(date 1524001094000)
+++ node_modules/fastcall/lib/native.js	(date 1524001094000)
@@ -16,4 +16,4 @@
 
 'use strict';
 
-module.exports = require('bindings')('fastcall');
\ No newline at end of file
+module.exports = require('../build/Release/fastcall');
\ No newline at end of file
