diff --git a/src/meson.build b/src/meson.build
index 34ccd8c..7678212 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -22,31 +22,22 @@ fc_sources = [
   'fcpat.c',
   'fcrange.c',
   'fcserialize.c',
   'fcstat.c',
   'fcstr.c',
   'fcweight.c',
   'fcxml.c',
   'ftglue.c',
 ]
 
-# FIXME: obviously fragile, cc.preprocess would be sweet
 cpp = cc.cmd_array()
-if cc.get_id() == 'gcc'
-  cpp += ['-E', '-P']
-elif cc.get_id() == 'msvc'
-  cpp += ['/EP']
-elif cc.get_id() == 'clang'
-  cpp += ['-E', '-P']
-else
-  error('FIXME: implement cc.preprocess')
-endif
+cpp += ['-E', '-P']
 
 cpp += ['-I', join_paths(meson.current_source_dir(), '..')]
 
 fcobjshash_gperf = custom_target('fcobjshash.gperf',
   input: 'fcobjshash.gperf.h',
   output: 'fcobjshash.gperf',
   command: [python3, files('cutout.py')[0], '@INPUT@', '@OUTPUT@', '@BUILD_ROOT@', cpp],
   build_by_default: true,
 )
 
